File: web/src/main/java/fr/arsenelapostolet/professor/server/core/entities/Student.java
package fr.arsenelapostolet.professor.server.core.entities;

import java.net.URI;
import java.util.Objects;
import java.util.Set;

public class Student {
    private final String id;
    private final String firstName;
    private final String lastName;
    private final String email;
    private final String gitlabUsername;
    private final Set<Grade> grades;
    private final String efreiClass;
    private final URI projectUrl;

    public Student(String id, String firstName, String lastName, String email, String gitlabUsername, Set<Grade> grades, String efreiClass, URI projectUrl) {
        this.id = id;
        this.firstName = firstName;
        this.lastName = lastName;
        this.email = email;
        this.gitlabUsername = gitlabUsername;
        this.grades = grades;
        this.efreiClass = efreiClass;
        this.projectUrl = projectUrl;
    }

    public String getId() {
        return id;
    }

    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public String getEmail() {
        return email;
    }

    public String getGitlabUsername() {
        return gitlabUsername;
    }

    public Set<Grade> getGrades() {
        return grades;
    }

    public String getEfreiClass() {
        return efreiClass;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Student student = (Student) o;
        return Objects.equals(id, student.id) && Objects.equals(email, student.email) && Objects.equals(gitlabUsername, student.gitlabUsername);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, email, gitlabUsername);
    }

    public URI getProjectUrl() {
        return this.projectUrl;
    }
}

File: app/src/main/kotlin/fr/arsenelapostolet/professor/core/entities/Student.kt
package fr.arsenelapostolet.professor.core.entities

import java.net.URI

class Student(
    val id: String,
    val firstName: String,
    val lastName: String,
    val email: String,
    val gitlabUsername: String,
    val grades: MutableList<Grade>,
    val efreiClass: String,
    val projectUrl: URI,
) {
    val fullName: String
        get() = "$firstName $lastName"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Student

        if (id != other.id) return false
        if (firstName != other.firstName) return false
        if (lastName != other.lastName) return false
        if (email != other.email) return false
        if (gitlabUsername != other.gitlabUsername) return false
        if (grades.size != other.grades.size) return false

        for (gradePair in grades.zip(other.grades)) {
            if (!gradePair.first.equals(gradePair.second)) {
                return false
            }
        }

        if (efreiClass != other.efreiClass) return false
        if (projectUrl != other.projectUrl) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + firstName.hashCode()
        result = 31 * result + lastName.hashCode()
        result = 31 * result + email.hashCode()
        result = 31 * result + gitlabUsername.hashCode()
        result = 31 * result + grades.hashCode()
        result = 31 * result + efreiClass.hashCode()
        result = 31 * result + projectUrl.hashCode()
        return result
    }


}